### DISCUSSION TITLE
Support for sh:sparql and sh:severity in shacl constraints

### DISCUSSION THREAD

**Discussion by logo021** [eclipse-rdf4j/rdf4j#4258] [Q&A] [answered]
I am learning shacl and is trying the examples from https://www.w3.org/TR/2017/REC-shacl-20170720/. When I try example "5.1 An Example SPARQL-based Constraint" it doesn't report a validation failure. I think it is because sh:sparql and related instructions, e.g. sh:SPARQLConstraint and sh:select are not implemented. 
Am I right and if so, is there any plans to implement it?
I also noted that sh:severity seem not be implemented, are there any plans to implement it as well?

**Comment by hmottestad** [comment, marked as answer]
Here is a list of our supported SHACL features: https://rdf4j.org/documentation/programming/shacl/#supported-shacl-features

We have both SPARQL based validation and support for severity in our plans. 

The current SHACL implementation is geared towards users who need to add constraints to a database and have the validation engine reject any transactions that would make the data in the database invalid. We use a novel approach to analyse the changes in a transaction to optimise the validation approach. This allows us to quickly validate a small change regardless of if you currently have a million or a billion triples in your database.

**Comment by logo021** [comment]
Hello Håvard

 

Do you know when SPARQL based validation and support for severity will be available?

 

In case it is far in future I will investigate other solutions.

 

Many thanks for responding / Lars-Ola

 

Från: Håvard M. Ottestad ***@***.***> 
Skickat: den 1 november 2022 09:32
Till: eclipse/rdf4j ***@***.***>
Kopia: Lars-Ola Osterlund ***@***.***>; Author ***@***.***>
Ämne: Re: [eclipse/rdf4j] Support for sh:sparql and sh:severity in shacl constraints (Discussion #4258)

 

Here is a list of our supported SHACL features: https://rdf4j.org/documentation/programming/shacl/#supported-shacl-features

We have both SPARQL based validation and support for severity in our plans.

The current SHACL implementation is geared towards users who need to add constraints to a database and have the validation engine reject any transactions that would make the data in the database invalid. We use a novel approach to analyse the changes in a transaction to optimise the validation approach. This allows us to quickly validate a small change regardless of if you currently have a million or a billion triples in your database.

**Reply by hmottestad** [reply]
At the moment we don’t have a specific date in mind. 

What is your use case and timeframe?

**Comment by logo021** [comment]
Currently I am investigating how to translate from OCL to SPARQL, see if I can do it with antlr. Will take to the end of this year. Once I have the SPARQL I will embed them in SHACL shapes, hope to begin next year.

 

Br / Lars-Ola

 

Från: Håvard M. Ottestad ***@***.***> 
Skickat: den 1 november 2022 17:18
Till: eclipse/rdf4j ***@***.***>
Kopia: Lars-Ola Osterlund ***@***.***>; Author ***@***.***>
Ämne: Re: [eclipse/rdf4j] Support for sh:sparql and sh:severity in shacl constraints (Discussion #4258)

 

At the moment we don’t have a specific date in mind.

What is your use case and timeframe?

**Reply by hmottestad** [reply]
That sounds interesting. Is it a part of a research project or a thesis? Or maybe a commercial use?

Are you going to use SPARQL exclusively or will you be using the other constraint components in SHACL?

**Comment by logo021** [comment]
I am just a retired guy interested in the topic. I have been working both with SHACL/SPARQL and EMF/OCL. I have a contact at KTH in Sweden and will suggest a paper on the topic if I find a conversion feasible. To convert OCL you will need SHACL for the logic and SPARQL for the model navigation.

 

Från: Håvard M. Ottestad ***@***.***> 
Skickat: den 5 november 2022 09:59
Till: eclipse/rdf4j ***@***.***>
Kopia: Lars-Ola Osterlund ***@***.***>; Author ***@***.***>
Ämne: Re: [eclipse/rdf4j] Support for sh:sparql and sh:severity in shacl constraints (Discussion #4258)

 

That sounds interesting. Is it a part of a research project or a thesis? Or maybe a commercial use?

Are you going to use SPARQL exclusively or will you be using the other constraint components in SHACL?

**Reply by hmottestad** [reply]
Maybe the SHACL Playground would be helpful during your development: https://shacl-playground.zazuko.com
